connection_pool/connector_test.go

```diff
@@ -0,0 +1,1168 @@
+	require.Equalf(t, reqResult, m.result, "unexpected result was passed")
+	require.Equalf(t, testMode, m.mode, "unexpected proxy mode")
+}
+
+type executeAsyncMock struct {
+	baseCallMock
+}
+
+func (m *executeAsyncMock) ExecuteAsync(functionName string, args interface{},
+	mode Mode) *tarantool.Future {
+	return m.callAsync(functionName, args, mode)
+}
+
+func TestConnectorExecuteAsync(t *testing.T) {
+	m := &executeAsyncMock{}
+	c := NewConnectorAdapter(m, testMode)
+
+	fut := c.ExecuteAsync(reqFunctionName, reqArgs)
+
+	require.Equalf(t, reqFuture, fut, "unexpected future")
+	require.Equalf(t, 1, m.called, "should be called only once")
+	require.Equalf(t, reqFunctionName, m.functionName,
+		"unexpected expr was passed")
+	require.Equalf(t, reqArgs, m.args, "unexpected args was passed")
+	require.Equalf(t, testMode, m.mode, "unexpected proxy mode")
+}
+
+var reqPrepared *tarantool.Prepared = &tarantool.Prepared{}
+
+type newPreparedMock struct {
+	Pooler
+	called int
+	expr   string
+	mode   Mode
+}
+
+func (m *newPreparedMock) NewPrepared(expr string,
+	mode Mode) (*tarantool.Prepared, error) {
+	m.called++
+	m.expr = expr
+	m.mode = mode
+	return reqPrepared, reqErr
+}
+
+func TestConnectorNewPrepared(t *testing.T) {
+	m := &newPreparedMock{}
+	c := NewConnectorAdapter(m, testMode)
+
+	p, err := c.NewPrepared(reqFunctionName)
+
+	require.Equalf(t, reqPrepared, p, "unexpected prepared")
+	require.Equalf(t, reqErr, err, "unexpected error")
+	require.Equalf(t, 1, m.called, "should be called only once")
+	require.Equalf(t, reqFunctionName, m.expr,
+		"unexpected expr was passed")
+	require.Equalf(t, testMode, m.mode, "unexpected proxy mode")
+}
+
+var reqStream *tarantool.Stream = &tarantool.Stream{}
+
+type newStreamMock struct {
+	Pooler
+	called int
+	mode   Mode
+}
+
+func (m *newStreamMock) NewStream(mode Mode) (*tarantool.Stream, error) {
+	m.called++
+	m.mode = mode
+	return reqStream, reqErr
+}
+
+func TestConnectorNewStream(t *testing.T) {
+	m := &newStreamMock{}
+	c := NewConnectorAdapter(m, testMode)
+
+	s, err := c.NewStream()
+
+	require.Equalf(t, reqStream, s, "unexpected stream")
+	require.Equalf(t, reqErr, err, "unexpected error")
+	require.Equalf(t, 1, m.called, "should be called only once")
+	require.Equalf(t, testMode, m.mode, "unexpected proxy mode")
+}
+
+var reqRequest tarantool.Request = tarantool.NewPingRequest()
+
+type doMock struct {
+	Pooler
+	called int
+	req    tarantool.Request
+	mode   Mode
+}
+
+func (m *doMock) Do(req tarantool.Request, mode Mode) *tarantool.Future {
+	m.called++
+	m.req = req
+	m.mode = mode
+	return reqFuture
+}
+
+func TestConnectorDo(t *testing.T) {
+	m := &doMock{}
+	c := NewConnectorAdapter(m, testMode)
+
+	fut := c.Do(reqRequest)
+
+	require.Equalf(t, reqFuture, fut, "unexpected future")
+	require.Equalf(t, 1, m.called, "should be called only once")
+	require.Equalf(t, reqRequest, m.req, "unexpected request")
+	require.Equalf(t, testMode, m.mode, "unexpected proxy mode")
+}
```

Comment:
I think it would be nice to have an example too.

Reply:
I've added ExampleConnectorAdapter.